wal/src/message_queue_impl/region.rs

```diff
@@ -0,0 +1,761 @@
+        log_topic: String,
+    ) -> Self {
+        // TODO: use snapshot to recover `region_meta`.
+        Self {
+            region_id,
+            region_meta: RegionMeta::default(),
+            log_encoding,
+            message_queue,
+            log_topic,
+        }
+    }
+
+    /// Write logs of table.
+    async fn write(
+        &self,
+        _ctx: &manager::WriteContext,
+        log_batch: &LogWriteBatch,
+    ) -> Result<SequenceNumber> {
+        // TODO: make use of ctx.
+        debug!(
+            "Begin to write to wal region, ctx:{:?}, region_id:{}, location:{:?}, log_entries_num:{}",
+            _ctx,
+            self.region_id,
+            log_batch.location,
+            log_batch.entries.len()
+        );
+        let location = &log_batch.location;
+        let log_write_entries = &log_batch.entries;
+
+        // Create messages and prepare for write.
+        let mut sequence_num = self
+            .region_meta
+            .prepare_for_table_write(location.table_id)
+            .await;
+        let mut messages = Vec::with_capacity(log_batch.entries.len());
+        let mut key_buf = BytesMut::new();
+        for entry in log_write_entries {
+            let log_key = CommonLogKey::new(self.region_id, location.table_id, sequence_num);
+            self.log_encoding
+                .encode_key(&mut key_buf, &log_key)
+                .map_err(|e| Box::new(e) as _)
+                .context(WriteLog {
+                    region_id: self.region_id,
+                    table_id: location.table_id,
+                })?;
+
+            let message = to_message(key_buf.to_vec(), entry.payload.clone());
+            messages.push(message);
+
+            sequence_num += 1;
+        }
+
+        // Write.
+        let offsets = self
+            .message_queue
+            .produce(&self.log_topic, messages)
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(WriteLog {
+                region_id: self.region_id,
+                table_id: location.table_id,
+            })?;
+        debug!(
+            "Produce to topic success, ctx:{:?}, region_id:{}, location:{:?}, topic:{}",
+            _ctx, self.region_id, log_batch.location, self.log_topic,
+        );
+
+        // Update after write.
+        self.region_meta
+            .update_after_table_write(
+                location.table_id,
+                OffsetRange::new(*offsets.first().unwrap(), *offsets.last().unwrap()),
+            )
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(WriteLog {
+                region_id: self.region_id,
+                table_id: location.table_id,
+            })?;
+
+        Ok(sequence_num - 1)
+    }
+
+    /// Read all logs from region, will return a iterator for high-level to
+    /// poll.
+    // TODO: take each read's timeout in consideration.
+    async fn read_all(
+        &self,
+        _ctx: &manager::ReadContext,
```

Comment:
```suggestion
        ctx: &manager::ReadContext,
```

Reply:
It is actually unused now, I will remove the `_` after making use of it.